jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushImageStep.java

```diff
@@ -40,84 +37,37 @@
 
+  // TODO: remove listeningExecutorService like other siblings
   PushImageStep(
       ListeningExecutorService listeningExecutorService,
       BuildConfiguration buildConfiguration,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
-      AuthenticatePushStep authenticatePushStep,
-      PushLayersStep pushBaseImageLayersStep,
-      PushLayersStep pushApplicationLayersStep,
-      PushContainerConfigurationStep pushContainerConfigurationStep,
-      BuildImageStep buildImageStep) {
+      Authorization pushAuthorization,
+      BlobDescriptor containerConfigurationDigestAndSize,
+      Image builtImage) {
     this.listeningExecutorService = listeningExecutorService;
     this.buildConfiguration = buildConfiguration;
     this.progressEventDispatcherFactory = progressEventDispatcherFactory;
-    this.authenticatePushStep = authenticatePushStep;
-    this.pushBaseImageLayersStep = pushBaseImageLayersStep;
-    this.pushApplicationLayersStep = pushApplicationLayersStep;
-    this.pushContainerConfigurationStep = pushContainerConfigurationStep;
-    this.buildImageStep = buildImageStep;
-
-    listenableFuture =
-        AsyncDependencies.using(listeningExecutorService)
-            .addStep(pushBaseImageLayersStep)
-            .addStep(pushApplicationLayersStep)
-            .addStep(pushContainerConfigurationStep)
-            .whenAllSucceed(this);
-  }
-
-  @Override
-  public ListenableFuture<BuildResult> getFuture() {
-    return listenableFuture;
+    this.pushAuthorization = pushAuthorization;
+    this.containerConfigurationDigestAndSize = containerConfigurationDigestAndSize;
+    this.builtImage = builtImage;
   }
 
   @Override
-  public BuildResult call() throws ExecutionException, InterruptedException {
-    return AsyncDependencies.using(listeningExecutorService)
-        .addStep(authenticatePushStep)
-        .addSteps(NonBlockingSteps.get(pushBaseImageLayersStep))
-        .addSteps(NonBlockingSteps.get(pushApplicationLayersStep))
-        .addStep(NonBlockingSteps.get(pushContainerConfigurationStep))
-        .addStep(NonBlockingSteps.get(buildImageStep))
-        .whenAllSucceed(this::afterPushSteps)
-        .get();
-  }
-
-  private BuildResult afterPushSteps() throws ExecutionException, InterruptedException {
-    AsyncDependencies dependencies = AsyncDependencies.using(listeningExecutorService);
-    for (AsyncStep<PushBlobStep> pushBaseImageLayerStep :
-        NonBlockingSteps.get(pushBaseImageLayersStep)) {
-      dependencies.addStep(NonBlockingSteps.get(pushBaseImageLayerStep));
-    }
-    for (AsyncStep<PushBlobStep> pushApplicationLayerStep :
-        NonBlockingSteps.get(pushApplicationLayersStep)) {
-      dependencies.addStep(NonBlockingSteps.get(pushApplicationLayerStep));
-    }
-    return dependencies
-        .addStep(NonBlockingSteps.get(NonBlockingSteps.get(pushContainerConfigurationStep)))
-        .whenAllSucceed(this::afterAllPushed)
-        .get();
-  }
```

Comment:
Confirming: the core logic (`afterAllPushed`) can run only after everything is concretely available:
1. push Authorization
1. base image layers downloaded and cached
1. application layers built and cached
1. target `Image` instance built
1. container configuration pushed into the target registry

Note that building a target `Image` instance already requires that all the layers are available.